beacon-chain/cache/depositcache/deposits_cache.go

```diff
@@ -84,17 +85,22 @@ func (dc *DepositCache) InsertDeposit(ctx context.Context, d *ethpb.Deposit, blo
 			"index":        index,
 			"deposit root": hex.EncodeToString(depositRoot[:]),
 		}).Warn("Ignoring nil deposit insertion")
-		return
+		return errors.New("nil deposit inserted into the cache")
 	}
 	dc.depositsLock.Lock()
 	defer dc.depositsLock.Unlock()
+
+	if int(index) != len(dc.deposits) {
+		return errors.Errorf("wanted deposit with index %d to be inserted but received %d", len(dc.deposits), index)
+	}
```

Comment:
If we only want to allow inserting deposits to the end, then the code that follows can be simplified to:
```
dc.deposits = append(dc.deposits, *dbpb.DepositContainer{Deposit: d, Eth1BlockHeight: blockNum, DepositRoot: depositRoot[:], Index: index})
historicalDepositsCount.Inc()
```

Reply:
This makes sense, but given how sensitive this part of the code is, I want to avoid any logic changes to deposit insertion in the PR which is why I have mostly focused on stricter validation.